### DISCUSSION TITLE
Переключатель срабатывания скрипта для веб страницы

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#118] [Помощь с алгоритмом] [answered]
Используя новую функцию добавления треков в очередь, создал скрипт вставляющий в очередь трек похожий на только что лайкнутый. Успешно протестировал на своем плейлисте дня, а так же в глобальном режиме при воспроизведении любого плейлиста. 
Т.к. реакция на текущее воспроизведение должна быть довольно быстрой, скрипт должен очень часто запускаться. Вроде пока справляется с триггером на 5 минут. Единственное, для исключения работы вхолостую, при работе в глобальном режиме, выставил условие запуска скрипта 
```
let playback = Player.getPlayback();
if (playback.isEmpty() || playback.context.type != 'playlist' || playback.repeat_state == 'track') {
  // прервать в ситуации когда нет данных, играет не плейлист, или повтор трека
  return;
}
```
В общем все отлично, но минус в том, что есть плейлисты где мне эта фишка нравится, а есть где лучше бы ее не использовать. Поэтому прошу помощи в реализации постоянного переключателя работы скрипта. Так же хочу вставить в условие плейлисты-исключения, в которых скрипт будет отрабатывать всегда, не зависимо от положения переключателя. Последнее через аддон не обязательно, достаточно прописать плейлисты-исключения в коде условия.

Заодно вопросик. 
В условии выставлена проверка, когда нет данных. Честно перекатал ее из очередного совета +18. Но что она означает? Я так понимаю, когда на сервере в плеере есть трек. Или играющий, или стоящий на паузе. При постановке на паузу, через какое-то время трек выгружается из плеера и это есть отсутствие данных. По моим наблюдениям этот момент наступает через 5 - 15 минут. 
А в свете последних нововведений, можно определить факт того что трек просто поставлен на паузу?

**Comment by Chimildic** [comment]
> что она означает?

Да, при некоторой продолжительной паузе Spotify не отдает такой трек. Кажется, даже если в приложении в плашке он остался. В таком случае `getPlayback` возвращает пустой объект `{}`. Можно потестировать, запуская вручную тест.
```js
function test(){
  console.log(Player.getPlayback());
}
```

> определить факт того что трек просто поставлен на паузу?

У плейбека есть булевый ключ `is_playing`. Когда играет `true`, когда на паузе `false`. Список всех ключей доступен при дебаге или в [документации](https://developer.spotify.com/documentation/web-api/reference/#endpoint-get-information-about-the-users-current-playback).
```js
let playback = Player.getPlayback();
if (playback.is_playing){
    // играет
} else {
    // пауза
}
```

> реализации постоянного переключателя

Получаются следующие режимы, которые будут переключатся на сайте?
- для всех плейлистов
- только избранных
- никаких

Можно полный код текущей функции и блоки сайта если есть.

**Reply by Bedmit** [reply]
Небольшое пояснение. WOW - это плейлист, куда при прослушивании скидывается очень-очень понравившийся трек. 

```
function LastLike() {
let playback = Player.getPlayback();
if (playback.isEmpty() || playback.context.type != 'playlist' || playback.repeat_state == 'track') {
  // прервать в ситуации когда нет данных, играет не плейлист, или повтор трека
  return;
}

let Past = Source.getSavedTracks();
let History = RecentTracks.get();

  let Wow = Source.getPlaylistTracks('WOW', '378MVstJa0dz36h0dYAu1N'); // получение и обработка интересного трека
  Selector.keepFirst(Wow, 1);
  let OthersTracks = Source.craftTracks(Wow);
  Order.sort(OthersTracks, 'meta.popularity', 'desc');
  
  Filter.matchExcept(OthersTracks, 'mix|club|radio|edit|live|karaoke');
  Filter.removeTracks(OthersTracks, Past);
  Filter.removeTracks(OthersTracks, History);
  Filter.rangeDateRel(History);
  Filter.removeArtists(OthersTracks, History);
  Filter.removeUnavailable(OthersTracks);
  Filter.dedupArtists(OthersTracks);
  Selector.keepFirst(OthersTracks, 3); // мини радио по интересному треку

  Filter.removeTracks(Wow, Wow);
  Playlist.saveWithReplace({id: '378MVstJa0dz36h0dYAu1N', name: 'WOW', public: false, tracks: Wow});

Selector.keepFirst(Past, 10);
let Last = Selector.sliceFirst(Past, 5);
let Like = Cache.read('LastLike.json');
Filter.removeTracks(Last, Like);
Cache.write('LastLike.json', Past);
Filter.replaceWithSimilar(Last, Source.getSavedTracks(), RecentTracks.get()); // замена лайков на похожие треки

Combiner.push(OthersTracks, Last); // мини радио + замена недавних лайков
Filter.dedupArtists(OthersTracks);
OthersTracks.forEach(t => Player.addToQueue(t)); // добавление в текущую очередь
}

```
Да, наверное три режима удобнее всего. 

И мне кажется кусок  `Filter.removeTracks(Wow, Wow);` неграмотный. Но я не знаю как по другому обнулить плейлист.

И как правильно оформить вместо проверки наличия данных, проверку на паузу. Не знаю как добавить `false`

**Comment by Chimildic** [comment, marked as answer]
> кажется кусок Filter.removeTracks(Wow, Wow); неграмотный

Указать пустой массив `[]` для `tracks` в `saveWithReplace`

Поскольку нет обратной связи со страницей, добавил пункт вывода текущего режима в консоль. 
```js
// launch.html
{
    action: 'toggleRecomQueueMode',
    title: 'Переключение режима',
    forms: [
        {
            key: 'mode',
            select: [
                { value: 'status', label: 'Узнать текущий режим' },
                { value: 'favorite', label: 'Только избранные плейлисты' },
                { value: 'overall', label: 'Любые плейлисты' },
                { value: 'disabled', label: 'Выключить рекомендации' },
            ]
        },
    ]
},
```

```js
// обработчик
function toggleRecomQueueMode(values) {
  const ACTION_ID = 'ActionRecomQueue';
  let state = Helper.readState(ACTION_ID);
  if (values.select.mode == 'status') {
    return 'Сейчас режим - ' + (state.mode || 'неопределен');
  } 
  state.mode = values.select.mode;
  Helper.updateState(ACTION_ID, state);
  return 'Режим переключен - ' + state.mode;
}
```
В константу `FAVORITE_IDS` через запятую добавлять _id_ избранных плейлистов. 
```js
function LastLike() {
  const FAVORITE_IDS = ['378MVstJa0dz36h0dYAu1N'];
  const FILENAME = 'LastLike.json';
  let state = Helper.readState('ActionRecomQueue');
  state.mode = state.mode || 'favorite';

  let playback = Player.getPlayback();
  if (state.mode == 'disabled'
    || playback.isEmpty()
    || !playback.is_playing
    || playback.repeat_state == 'track'
    || playback.context.type != 'playlist'
    || (state.mode == 'favorite' && !FAVORITE_IDS.includes(Helper.parseId(playback.context.uri)))) {
    // прервать в ситуации когда нет данных, трек на паузе или повторе, играет не плейлист, не соответствие режима
    return;
  }

  let savedTracks = Source.getSavedTracks();
  let history = RecentTracks.get();

  let wowTracks = Source.getPlaylistTracks('WOW', '378MVstJa0dz36h0dYAu1N'); // получение и обработка интересного трека
  Selector.keepFirst(wowTracks, 1);
  let craftedTracks = Source.craftTracks(wowTracks);
  Order.sort(craftedTracks, 'meta.popularity', 'desc');

  Filter.matchExcept(craftedTracks, 'mix|club|radio|edit|live|karaoke');
  Filter.removeTracks(craftedTracks, savedTracks);
  Filter.removeTracks(craftedTracks, history);
  Filter.rangeDateRel(history);
  Filter.removeArtists(craftedTracks, history);
  Filter.removeUnavailable(craftedTracks);
  Filter.dedupArtists(craftedTracks);
  Selector.keepFirst(craftedTracks, 3); // мини радио по интересному треку

  Playlist.saveWithReplace({ id: '378MVstJa0dz36h0dYAu1N', name: 'WOW', public: false, tracks: [] });

  Selector.keepFirst(savedTracks, 10);
  let recentLikes = Selector.sliceFirst(savedTracks, 5);
  let cacheLike = Cache.read(FILENAME);
  Filter.removeTracks(recentLikes, cacheLike);
  Cache.write(FILENAME, savedTracks);
  Filter.replaceWithSimilar(recentLikes, Source.getSavedTracks(), RecentTracks.get()); // замена лайков на похожие треки

  Combiner.push(craftedTracks, recentLikes); // мини радио + замена недавних лайков
  Filter.dedupArtists(craftedTracks);
  craftedTracks.forEach(t => Player.addToQueue(t)); // добавление в текущую очередь
}
```

**Reply by Bedmit** [reply]
С основным скриптом все понял. Первые две части, не очень.
Как вставить код в launch.html , к остальным блокам? Или это вообще как-то отдельно будет?
```
const func = [
                {
                  name: 'RadioByTrack',
                  text: 'Радио по треку',
                  args: [{ name: 'Ссылка на трек',key: 'strTrackId' }],
                },
                {
                  name: 'Fresh',
                  text: 'Освежить плейлист',
                  args: [{ name: 'Ссылка на плейлист', key: 'playlistUrl' }],
                }, 
                {
                  name: 'runFunction',
                  text: 'Обновить плейлист',
                  args: [
                {
                  key: 'name',
                  select: [
                { value: 'History', name: 'History' },    
                { value: 'Bedtime', name: 'Bedtime' },
                { value: 'BigSmartify', name: 'BigSmartify' },
                ],
                },
                ],
                },
            ];

```
Обработчик - это отдельная функция? Ее можно например и в скрипте после `function LastLike()` вставить?

**Reply by Chimildic** [reply]
Как обычно настраиваются блоки для сайта. Первый к перечню остальных блоков в константе из `launch.html`, второй в файл `main` например. Или да, после `LastLike`.

Судя по примеру `const func`, остались старые имена. В начале [темы](https://github.com/Chimildic/goofy/discussions/9) есть пояснение про новые `ACTIONS`.

**Reply by Bedmit** [reply]
Точно, обновил `launch.html` и все получилось. Правда старые блоки поломались. Ну да не страшно, сейчас переделаю.

Очередной раз спасибо! Сам бы такое точно не сделал. Мои идеи и сложность их реализации из-за нехватки знаний, все больше и больше подталкивают к хотя бы базовому изучению языка.

А скрипт кстати, с этой новой возможностью добавления в очередь, просто преотличный! Похвастаюсь в клубе пожалуй ))
